speechbrain/decoders/scorer.py

```diff
@@ -0,0 +1,850 @@
+    ... )
+    >>> coverage_scorer = CoverageScorer(vocab_size=vocab_size)
+    >>> scorer = ScorerBuilder(
+    ...     full_scorers=[rnnlm_scorer, coverage_scorer],
+    ...     weights={'rnnlm': lm_weight, 'coverage': coverage_penalty}
+    ... )
+    >>> beam_size=5
+    >>> searcher = S2SRNNBeamSearcher(
+    ...     embedding=emb,
+    ...     decoder=dec,
+    ...     linear=seq_lin,
+    ...     bos_index=1,
+    ...     eos_index=2,
+    ...     min_decode_ratio=0.0,
+    ...     max_decode_ratio=1.0,
+    ...     topk=2,
+    ...     using_eos_threshold=False,
+    ...     beam_size=beam_size,
+    ...     temperature=1.25,
+    ...     scorer=scorer
+    ... )
+    >>> batch_size=2
+    >>> enc = torch.rand([batch_size, n_channels, d_model])
+    >>> wav_len = torch.ones([batch_size])
+    >>> topk_hyps, topk_lengths, topk_scores, topk_log_probs = searcher(enc, wav_len)
+    """
+
+    def __init__(self, vocab_size, threshold=0.5):
+        self.vocab_size = vocab_size
+        self.threshold = threshold
+        # Use time_step to normalize the coverage over steps
+        self.time_step = 0
+
+    def score(self, inp_tokens, coverage, candidates, attn):
+        """Specifies token scoring."""
+        n_bh = attn.size(0)
+        self.time_step += 1
+
+        if coverage is None:
+            coverage = torch.zeros_like(attn, device=attn.device)
+
+        # Current coverage
+        if len(attn.size()) > 2:
+            # the attn of transformer is [batch_size x beam_size, current_step, source_len]
+            coverage = torch.sum(attn, dim=1)
+        else:
+            coverage = coverage + attn
+
+        # Compute coverage penalty and add it to scores
+        penalty = torch.max(
+            coverage, coverage.clone().fill_(self.threshold)
+        ).sum(-1)
+        penalty = penalty - coverage.size(-1) * self.threshold
+        penalty = penalty.view(n_bh).unsqueeze(1).expand(-1, self.vocab_size)
+        return -1 * penalty / self.time_step, coverage
+
+    def permute_mem(self, coverage, index):
+        """Specifies memory synchronisation."""
+        # Update coverage
+        coverage = torch.index_select(coverage, dim=0, index=index)
+        return coverage
+
+    def reset_mem(self, x, enc_lens):
+        """Specifies memory reset."""
+        self.time_step = 0
+        return None
+
+
+class LengthScorer(BaseScorerInterface):
+    """A length rewarding scorer.
+
+    Arguments
+    ---------
+    vocab_size: int
+        The total number of tokens.
+    """
+
+    def __init__(self, vocab_size):
+        self.vocab_size = vocab_size
+
+    def score(self, inp_tokens, memory, candidates, attn):
+        """Specifies token scoring."""
+        return (
+            torch.tensor(
+                [1.0], device=inp_tokens.device, dtype=inp_tokens.dtype
+            ).expand(inp_tokens.size(0), self.vocab_size),
+            None,
+        )
+
+
+class ScorerBuilder:
+    """ Builds scorer instance for beamsearch.
```

Comment:
More docstring, what is this usefull for?

Reply:
done